chain/src/main/scala/org/bitcoins/chain/blockchain/ChainHandler.scala

```diff
@@ -92,24 +115,168 @@ case class ChainHandler(
     }
     Future.successful(hashBE)
   }
+
+  override def nextBatchRange(prevStopHash: DoubleSha256DigestBE, batchSize: Long)(implicit ec: ExecutionContext): Future[Option[(Int, DoubleSha256Digest)]] = {
+    val startHeightF = if (prevStopHash == DoubleSha256DigestBE.empty) {
+      Future.successful(0)
+    } else {
+      for {
+        prevStopHeaderOpt <- getHeader(prevStopHash)
+        prevStopHeader = prevStopHeaderOpt.getOrElse(throw new RuntimeException(s"Unknown block hash ${prevStopHash}"))
+      } yield prevStopHeader.height + 1
+    }
+    for {
+      startHeight <- startHeightF
+      blockCount <- getBlockCount
+      stopHeight = if (startHeight - 1 + batchSize > blockCount) blockCount else startHeight - 1 + batchSize
+      stopBlockOpt <- getHeadersByHeight(stopHeight.toInt).map(_.headOption)
+      stopBlock = stopBlockOpt.getOrElse(throw new RuntimeException(s"Unknown header height ${stopHeight}"))
+    } yield {
+      if (startHeight > stopHeight)
+        None
+      else
+        Some((startHeight, stopBlock.hashBE.flip))
+    }
+  }
+
+  override def processFilterHeaders(
+      filterHeaders: Vector[FilterHeader],
+      stopHash: DoubleSha256DigestBE)(
+      implicit ec: ExecutionContext): Future[ChainApi] = {
+
+    val filterHeadersToCreateF = for {
+      blockHeaders <- blockHeaderDAO.getNChildren(stopHash, filterHeaders.size - 1).map(_.sortBy(_.height))
+    } yield {
+      if (blockHeaders.size != filterHeaders.size) {
+        throw new RuntimeException(s"Filter header batch size does not match block header batch size ${filterHeaders.size} != ${blockHeaders.size}")
+      }
+      blockHeaders.indices.toVector.map { i =>
+        val blockHeader = blockHeaders(i)
+        val filterHeader = filterHeaders(i)
+        CompactFilterHeaderDbHelper.fromFilterHeader(
+          filterHeader,
+          blockHeader.hashBE,
+          blockHeader.height)
+      }
+    }
+
+    for {
+      filterHeadersToCreate <- filterHeadersToCreateF
+      _ <- if (filterHeadersToCreate.nonEmpty && filterHeadersToCreate.head.height > 0) {
+        filterHeaderDAO.findByHash(filterHeadersToCreate.head.previousFilterHeaderBE).map { prevHeaderOpt =>
+          require(prevHeaderOpt.nonEmpty, s"Previous filter header does not exist: ${filterHeadersToCreate.head.previousFilterHeaderBE}")
+          require(prevHeaderOpt.get.height == filterHeadersToCreate.head.height - 1, s"Unexpected previous header's height: ${prevHeaderOpt.get.height} != ${filterHeadersToCreate.head.height - 1}")
+        }
+      } else Future.unit
+      _ <- filterHeaderDAO.createAll(filterHeadersToCreate)
+    } yield this
+  }
+
+  override def processFilters(messages: Vector[CompactFilterMessage])(implicit ec: ExecutionContext): Future[ChainApi] = {
+    for {
+      filterHeaders <- filterHeaderDAO.findAllByBlockHashes(messages.map(_.blockHash.flip)).map(_.sortBy(_.height))
```

Comment:
so a couple thoughts in general with `processFilters()` and `processFilterHeaders()`

1. It seems like the control flow is very similar for these methods. Can we make a private helper method and pass in functions to do the custom stuff (for instance, inserting into `filterHeaderDAO` vs `filterDAO`. 
2. Putting everything in a for expression and _not_ using intermediate vals makes this code really hard to read. This really needs to be broken up so it can be reviewed. 

Something like this is much better:

```scala
  override def processFilters(messages: Vector[CompactFilterMessage])(
      implicit ec: ExecutionContext): Future[ChainApi] = {

    val filterHeadersF = filterHeaderDAO
      .findAllByBlockHashes(messages.map(_.blockHash.flip))
      .map(_.sortBy(_.height))

    val messagesByBlockHash = messages.groupBy(_.blockHash.flip)

    val sizeCheckF = for {
      filterHeaders <- filterHeadersF
      _ <- if (filterHeaders.length != messagesByBlockHash.last) {
        Future.failed(new RuntimeException(
          s"Filter batch size does not match filter header batch size ${messages.size} != ${filterHeaders.size}"))
      } else {
        FutureUtil.unit
      }
    } yield ()

    for {
      filterHeaders <- filterHeadersF
      _ <- sizeCheckF
      compactFilterDbs = filterHeaders.map { filterHeader =>
        findFilterDbFromMessage(filterHeader, messagesByBlockHash)
      }
      _ <- filterDAO.createAll(compactFilterDbs)
    } yield {
      this
    }
  }

  private def findFilterDbFromMessage(
      filterHeader: CompactFilterHeaderDb,
      messagesByBlockHash: Map[
        DoubleSha256DigestBE,
        Vector[CompactFilterMessage]]): CompactFilterDb = {
    messagesByBlockHash.get(filterHeader.blockHashBE) match {
      case Some(messages) if messages.size == 1 =>
        val message = messages.head
        val filterHashBE = CryptoUtil.doubleSHA256(message.filterBytes).flip
        if (filterHashBE != filterHeader.filterHashBE) {
          //shouldn't we be throwing here? This seems really bad!
          val errMsg = s"Filter hash does not match filter header hash: ${filterHashBE} != ${filterHeader.filterHashBE}\n" +
            s"filter=${message.filterBytes.toHex}\nblock hash=${message.blockHash}\nfilterHeader=${filterHeader}"
          logger.warn(errMsg)
        }
        val filter =
          CompactFilterDbHelper.fromFilterBytes(message.filterBytes,
                                                filterHeader.blockHashBE,
                                                filterHeader.height)
        filter
      case error @ _ =>
        //yikes! what does this mean?
        throw new RuntimeException()
    }
  }
```

Reply:
It seems to me the only similarity between those two methods is `createAll()`